Modified coeff() to give correct value for a sum

t-ens opened the pull request:
#### References to other Issues or PRs
Fixes #16752 
Fixes #17917

#### Brief description of what is fixed or changed
Given expressions e and n which are both noncommutative when computing e.coeff(n) instead of immediately returning 0 if n is found in more than one summand of e it now only returns 0 when finding n in a new summand if, after removing n from both summands, the remaining noncommutative part of the new summand  and previous summand are not equal.  Added tests for this behaviour as well.

#### Other comments
Before if everything was commutative then for example `(x*a+y*a+z*b).coeff(a)` returned `x+y` and even if `a,b` were not commutative `(x*a+y*a).coeff(a)` returned `x+y` however for `a,b` not commutative `(x*a+y*a+z*b).coeff(a)` returned `0`.  With this change now `(x*a+y*a+z*b).coeff(a)` returns `x+y` in all cases. This fixes the incorrect answers given by `LieDerivative` and `Commutator` from the diffgeom submodule referenced above however it is slower so there is a performance hit.

<!-- BEGIN RELEASE NOTES -->

* core
  * Extended coeff() to return nonzero value for some sums with noncommutative variables
 
<!-- END RELEASE NOTES -->

t-ens committed:
Modified coeff() to give correct value for a sum

For example given  x*a+y*a+z*b where a,b are not commutative coeff() now returns
x+y rather than 0

oscarbenjamin commented:
The tests failed because of the changes to the authors file. I'm about to disable them in #21987. Once that is merged you can close and reopen the PR to rerun the tests.

smichr commented:
cf #11196

oscargus commented on sympy/core/expr.py:
```diff
@@ -1619,22 +1619,31 @@ def find(l, sub, first=True):
-            else:
-                if hit:
-                    ii, r, n = hit
+                        if not right:
```
Minor thing, but I'd swap the order of these if-else clauses so that the negation is avoided.

t-ens replied:
I did it this way because whoever wrote the function originally wrote if-else clauses with negation(for example line 1594 and a few other places.) I fixed my own contribution to avoid the negation. Should I also change the other occurrences of this in the coeff() function which I did not write?

oscargus replied:
I think it is enough to change it here (although not required). Unless there is a "pedagogical" benefit of starting with the not-clause it is marginally better not to. But there is indeed a lot of code where it is not like that. I actually tried the other day to fix it systematically, but in some parts of the code base, it really needs to be written with not first...

oscargus commented on sympy/core/tests/test_expr.py:
```diff
@@ -1414,6 +1414,8 @@ def test_coeff():
     assert (n*m + o*m*n).coeff(m*n, right=1) == 1
     assert (n*m + n*m*n).coeff(n*m, right=1) == 1 + n  # = n*m*(n + 1)
+    assert (x*n + y*n + z*m).coeff(n) == x+y
+    assert (n*m +n*o + o*l).coeff(n,right=1) == m+o
```
Spaces around +

t-ens replied:
Oops, just noticed there should also be a space before right in the second one.  I'll commit this together with the other if-else clause changes if I need to do those.

t-ens committed:
Coding style fixes

oscargus commented:
Oh, I didn't mention that, but it really makes sense to add tests for the issues you close. For #16752 it is probably enough to add a comment where you test x*a + y*a + z*b like `# issue #16752`, but for #17917 please add a test `test_issue_17917` in https://github.com/sympy/sympy/blob/master/sympy/diffgeom/tests/test_diffgeom.py

t-ens committed:
Added test for issue 17917 and more coding style fixes

t-ens commented:
After testing the function a bit more I realized that even though it passed all the tests my change is not correct. For example if `e=xuvu+yuvw+zw` my code gives `e.coeff(v, right=True)=u+w` because I only checked whether the noncommutative parts to the left of `v` are equal not the commutative parts and then added the coefficient from each summand. This is incorrect. If I simply  ensure that both the noncommutative part of the expression to the left of `v` and the commutative parts are equal, then I would get `e.coeff(v, right=True)=0` which is also incorrect since `xuvu+yuvw=uv(xu+yw)` so the correct answer should be `xu+yw`.

On the other hand using the original code before I modified anything we correctly get `(xuvu+yuvw).coeff(v, right=True)=xu+yw` as we should, however now `e.coeff(v,right=True) = 0`. The problem is because of the extra term `zw` which does not contain `v` at all which causes the tests for special cases where there is a common factor in the noncommutative part of each term to fail. What I really need to do is remove all terms which do not have `v` at all, and then use the algorithms already implemented on the remaining terms which do more complicated factoring than I was doing.

So the most obvious solution would be to simply revert my changes back to the way they were and add the line
`co = [t for t in co if find(t[1], nx, right) is not None]`
in line 1589. I've tested this and it solves the problems with the issues in the PR, passes all the tests and also gives the right answer for the above example.

Is this an acceptable solution? I ask because it seems like it is quite inefficient to apply the `find()` function to every term right away like this  since the special cases sometimes only require you to apply this function to a single term so for large expressions containing many terms adding my above line could cause a significant slowdown. One way I see to only have the slowdown when it is actually required would be to add an argument to the coeff() function, say `reduced` and have it be false by default. When `e.coeff(n)` is called the function would run to line 1622. If the all the special cases fail and `reduced` is False then use my line above to remove any term not containing `n` and then call coeff() again this time with `reduced=True` and the new expression so that now maybe the special cases apply now that any terms not containing `n` have been removed. If all the special cases fail, now reduced is True so now search for a single match.

Also since I am essentially changing my pull request completely, is it best to close this pull request and start a new one or just add another commit to this one? Sorry, I'm new to this and not really sure what the best way to proceed is.

t-ens committed:
coeff() now removes all terms not containing a match before trying to extract coefficients again if necessary. Added another test for an example coeff() was silently getting wrong.

t-ens commented:
I have added the requested test and addressed all the other issues that were still relevant with my new update. As per my previous post I changed approach in my latest commit since my previous algorithm did not work in all cases. Rather than running the line
` co = [t for t in co if find(t[1], nx, right) is not None]`
for every expression I thought it would be more efficient to first try the special cases, then only run this line if necessary since the special cases only require performing the find() one time rather than once for every term in the expression. I also added a test for the example in the previous post which was incorrect in both my updated version and the original version of the function.

I'm not sure if I should close this pull request and make a new one or not since I changed my approach completely and reverted most of my original code changes. Let me know if I should or if there is a problem with the new approach.

oscargus commented:
I think that you should continue in the same PR.

A bit strange that it fails in 3.10 and passes in the other versions...

oscarbenjamin commented:
> A bit strange that it fails in 3.10 and passes in the other versions...

There is an intermittent failure in 3.10 in CI (#21968).

t-ens commented:
I believe the issue with the test was fixed in #22029 so closed and reopened to rerun the tests and also removed the draft status since I think it works correctly now.

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1414,6 +1414,9 @@ def test_coeff():
     assert (n*m + n*m*n).coeff(n*m, right=1) == 1 + n  # = n*m*(n + 1)
+    assert (x*n + y*n + z*m).coeff(n) == x + y #i issue #16752
+    assert (n*m + n*o + o*l).coeff(n, right=1) == m + o
+    assert (x*n*m*n + y*n*m*o + z*l).coeff(m, right=1) == x*n + y*o
```
I think we need an assurance the coefficient of the left and right will recreate the terms that were matched:

```
>>> (x*n*m*n + y*n*m*o + z*l).coeff(m, right=0)
n
>>> (x*n*m*n + y*n*m*o + z*l).coeff(m, right=1)
x*n + y*o
```
That's right because `factor_nc(x*n*m*n + y*n*m*o) == n*m*(x*n + y*o)`. But consider the following:
```python
>>> (x*n*m*n + x*n*m*o + z*l).coeff(m, right=0)
x*n
>>> (x*n*m*n + x*n*m*o + z*l).coeff(m, right=1)
x*n + x*o
```
The `x` cannot be part of the left and part of the right.

smichr commented:
Doesn't this single change accomplish the same thing as this proposal?
From this
```python
    if self_c and not x_c:
        return S.Zero
```
to this:
```python
    if self_c and not x_c:
        return S.Zero
    if self.is_Add and not self_c and not x_c:
        d = self.as_independent(x)[1]
        return coeff(d, x, right=right)
```
The issue of returning the commutative part as both left and right coeff still remains, however, and is present in master.

smichr commented:
Am trying alternative in #22137

smichr commented:
@t-ens , what do you think about #22137?

t-ens commented:
I agree that if `e` is some expression containing `x` and `a`, `b` are the left and right coefficients of `x` in `e` then it is reasonable to expect that `e = axb`. #22140 looks like a better implementation of what I was trying to do here plus it addresses this reconstruction issue as well so this pull request can be closed in favour of that one.
